Sources/Sentry/SentryHttpInterceptor.m

```diff
@@ -0,0 +1,112 @@
+        ? [NSMutableArray arrayWithArray:[configuration protocolClasses]]
+        : [[NSMutableArray alloc] init];
+
+    if (![protocolClasses containsObject:[self class]]) {
+        [protocolClasses insertObject:[self class] atIndex:0];
+    }
+
+    configuration.protocolClasses = protocolClasses;
+}
+
++ (BOOL)canInitWithRequest:(NSURLRequest *)request
+{
+    // Intercept the request if it is a http/https request
+    // not targeting Sentry and there is transaction in the scope.
+    NSNumber *intercepted = [NSURLProtocol propertyForKey:SENTRY_INTERCEPTED_REQUEST
+                                                inRequest:request];
+    if (intercepted != nil && [intercepted boolValue])
+        return false;
+
+    NSURL *apiUrl = [NSURL URLWithString:SentrySDK.options.dsn];
+    if ([request.URL.host isEqualToString:apiUrl.host] &&
+        [request.URL.path containsString:apiUrl.path])
+        return false;
+
+    if (SentrySDK.currentHub.scope.span == nil)
+        return false;
+
+    return ([request.URL.scheme isEqualToString:@"http"] ||
+        [request.URL.scheme isEqualToString:@"https"]);
+}
+
++ (NSURLRequest *)canonicalRequestForRequest:(NSURLRequest *)request
+{
+    id<SentrySpan> span = SentrySDK.currentHub.scope.span;
+    if (span == nil)
+        return request;
+
+    NSMutableURLRequest *newRequest = [request mutableCopy];
+    [newRequest addValue:[span toTraceHeader].value forHTTPHeaderField:SENTRY_TRACE_HEADER];
+    [NSURLProtocol setProperty:@YES forKey:SENTRY_INTERCEPTED_REQUEST inRequest:newRequest];
+    return newRequest;
+}
+
+- (void)startLoading
+{
+    NSURLSessionConfiguration *conf = [NSURLSessionConfiguration defaultSessionConfiguration];
+    self.session = [NSURLSession sessionWithConfiguration:conf delegate:self delegateQueue:nil];
+    [[self.session dataTaskWithRequest:self.request] resume];
+}
+
+- (void)stopLoading
+{
+    [self.session invalidateAndCancel];
+    self.session = nil;
+}
+
+#pragma mark - NSURLSession Delegate
+
+- (void)URLSession:(NSURLSession *)session
+                    task:(NSURLSessionTask *)task
+    didCompleteWithError:(NSError *)error
+{
+    if (error) {
+        [self.client URLProtocol:self didFailWithError:error];
+    } else {
+        [self.client URLProtocolDidFinishLoading:self];
+    }
+}
+
+- (void)URLSession:(NSURLSession *)session
+              dataTask:(NSURLSessionDataTask *)dataTask
+    didReceiveResponse:(NSURLResponse *)response
+     completionHandler:(void (^)(NSURLSessionResponseDisposition))completionHandler
+{
+    [self.client URLProtocol:self
+          didReceiveResponse:response
+          cacheStoragePolicy:NSURLCacheStorageNotAllowed];
+
+    completionHandler(NSURLSessionResponseAllow);
+}
+
+- (void)URLSession:(NSURLSession *)session
+          dataTask:(NSURLSessionDataTask *)dataTask
+    didReceiveData:(NSData *)data
+{
+    [self.client URLProtocol:self didLoadData:data];
+}
```

Comment:
`l`: I'm not an expert on `NSURLProtocol`. Can you tell me why we need all that?

Reply:
With or custom NSURLProtocol we are telling the system that we will handle the request. 
Since that is no point in implementing the HTTP protocol, we create another URLSession that will not be intercepted and use it as a proxy.

Reply:
Since this code impacts basically all HTTP requests, I think we must add tests for that logic. Even if we know it is working now, we can easily break our users with a small modification.

Reply:
Excellent point, this could potentially mess user's entire project.
I will work on these tests.

Reply:
Tests created!!